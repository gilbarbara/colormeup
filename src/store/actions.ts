export { alertHide, alertShow } from './slices/alerts';
export { toggleSidebar } from './slices/app';
export { setColor, setOptions } from './slices/color';
export { removeColor, resetUserData, saveColor, setUserOptions } from './slices/user';
