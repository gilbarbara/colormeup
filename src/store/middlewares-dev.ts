import { Middleware } from '@reduxjs/toolkit';
import invariant from 'redux-immutable-state-invariant';
import { createLogger } from 'redux-logger';

export default [invariant(), createLogger({ collapsed: true })] as Middleware[];
