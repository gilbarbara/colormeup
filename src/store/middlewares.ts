import { Middleware } from 'redux';

const middlewares: Middleware[] = [];

if (process.env.NODE_ENV === 'development') {
  // eslint-disable-next-line @typescript-eslint/no-require-imports
  const devMiddlewares = require('./middlewares-dev').default;

  middlewares.push(...devMiddlewares);
}

export default middlewares;
