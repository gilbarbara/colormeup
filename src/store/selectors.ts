import { createSelector } from '@reduxjs/toolkit';

import { RootState } from '~/types';

const selectSelf = (state: RootState) => state;

export const selectApp = createSelector(selectSelf, state => state.app);

export const selectColor = createSelector(selectSelf, state => state.color);

export const selectUser = createSelector(selectSelf, state => state.user);
