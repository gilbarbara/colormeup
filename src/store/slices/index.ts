export { alertHide, alertShow } from './alerts';
export { toggleSidebar } from './app';
export { setColor, setOptions } from './color';
export { removeColor, resetUserData, saveColor, setUserOptions } from './user';
