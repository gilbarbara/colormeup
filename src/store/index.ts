import { configureStore, Tuple } from '@reduxjs/toolkit';
import {
  FLUSH,
  PAUSE,
  PERSIST,
  persistCombineReducers,
  persistStore,
  PURGE,
  REGISTER,
  REHYDRATE,
} from 'redux-persist';
import autoMergeLevel2 from 'redux-persist/lib/stateReconciler/autoMergeLevel2';
import storage from 'redux-persist/lib/storage';

import { RootState } from '~/types';

import dynamicMiddlewares from './dynamic-middlewares';
import middlewares from './middlewares';
import alerts, { alertsState } from './slices/alerts';
import app, { appState } from './slices/app';
import color, { colorState } from './slices/color';
import user, { userState } from './slices/user';

const getDefaultMiddlewareOptions = {
  serializableCheck: {
    ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
  },
  thunk: false,
};

export const initialState = {
  alerts: alertsState,
  app: appState,
  color: colorState,
  user: userState,
};

export const reducers = {
  alerts,
  app,
  color,
  user,
};

const rootReducer = persistCombineReducers<RootState>(
  {
    key: 'colormeup',
    stateReconciler: autoMergeLevel2,
    storage,
    blacklist: ['alerts'],
    timeout: 0,
    version: 1,
  },
  reducers,
);

/* istanbul ignore next */
export const configStore = (preloadedState: any = {}) => {
  const storeEnhanced = configureStore({
    reducer: rootReducer,
    middleware: getDefaultMiddleware => {
      return new Tuple(
        ...getDefaultMiddleware(getDefaultMiddlewareOptions),
        ...middlewares,
        dynamicMiddlewares,
      );
    },
    preloadedState,
  });

  return {
    persistor: persistStore(storeEnhanced),
    store: storeEnhanced,
  };
};

export const { persistor, store } = configStore();
