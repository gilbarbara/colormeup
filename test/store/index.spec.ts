import { persistor, store } from '~/store';

vi.mock('@gilbarbara/helpers', () => {
  const helpers = vi.importActual('@gilbarbara/helpers');

  return {
    ...helpers,
    now: () => 1234567890,
  };
});

describe('store', () => {
  it('should have a store', () => {
    expect(store.getState()).toMatchSnapshot();
  });

  it('should have a persistor', () => {
    expect(persistor.getState()).toEqual({
      bootstrapped: true,
      registry: [],
    });
  });
});
