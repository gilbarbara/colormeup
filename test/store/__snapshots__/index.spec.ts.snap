// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`store > should have a store 1`] = `
{
  "_persist": {
    "rehydrated": true,
    "version": 1,
  },
  "alerts": {
    "data": [],
  },
  "app": {
    "isSidebarOpen": false,
  },
  "color": {
    "hex": "#ff0044",
    "model": "hsl",
    "steps": 16,
    "type": "h",
  },
  "user": {
    "colors": [],
    "createdAt": 1234567890,
    "showHelp": true,
    "updatedAt": 1234567890,
  },
}
`;
